src/NLog/LogEventInfo.cs

```diff
@@ -523,6 +536,31 @@ private void CalcFormattedMessage()
             }
         }
 
+        private bool ResetFormattedProperties()
+        {
+            if (this.properties != null && this.properties.Count > 0 && this.logMessageFormatter.HasProperties(this))
+            {
+                for (int i = 0; i < this.parameters.Length; ++i)
+                {
+                    object parameterName = null;
+                    foreach (var property in this.properties)
```

Comment:
I have to agree this is really the downside of the same collection!

What do think about: internally split it to 2 collections (messageTemplateParameters, properties), and expose it as one collection to the outside?